### PR TITLE
feat: replace jwt decode and roles to use fam authorization on be

### LINKED ISSUE #481: Replace JWT decode and roles to use FAM authorization on BE

### PR DESCRIPTION
**Pull request by RMCampos**
# Description

Updates both `backend` and `oracle-api` to be able to decode FAM Cognito token.

Fixes #481 

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Strongly tested locally.

## Checklist

- [x] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have already been accepted and merged

## Further comments

---
Thanks for the PR!

Any successful deployments (not always required) will be available below.
[Backend](https://nr-spar-585-backend.apps.silver.devops.gov.bc.ca/)
[Frontend](https://nr-spar-585-frontend.apps.silver.devops.gov.bc.ca/)
[Oracle-API](https://nr-spar-585-oracle-api.apps.silver.devops.gov.bc.ca/)

Once merged, code will be promoted and handed off to following workflow run.
[Main Merge Workflow](https://github.com/bcgov/nr-spar/actions/workflows/merge-main.yml)

### CODE REVIEW

File: backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java [hunk 1]
```diff
@@ -31,42 +26,58 @@ public Optional<UserInfo> getUserInfo() {
...
-                jwtPrincipal.getClaimAsString("bceid_business_name"),
-                IdentityProvider.fromClaim(jwtPrincipal).orElseThrow(),
+                idpUsername,
+                firstName,
+                lastName,
+                email,
+                displayName,
+                isIdirProvider ? idpUsername : null,
+                isIdirProvider ? null : idpUsername,
+                IdentityProvider.fromClaim(provider).orElseThrow(),
```
Comment (craigyu) [inline backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java:80]: can we have the userId in here? i think the `idpUsername` does not contain the `@idir` or `@bceid` string at the end.

File: backend/src/test/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelperTest.java [hunk 2]
```diff
@@ -40,26 +36,24 @@ void getUserInfoIdirTest() {
...
+    builder.subject("BAGGINGS");
     builder.header("alg", "HS256");
     builder.header("typ", "JWT");
-    builder.claim("given_name", "Bilbo");
-    builder.claim("family_name", "Baggings");
     builder.claim("email", "bilbo.baggings@gov.bc.ca");
-    builder.claim("display_name", "Baggings, Bilbo LWRS:EX");
-    builder.claim("idir_username", "BAGGINGS");
-    builder.claim("identity_provider", "idir");
+    builder.claim("custom:idp_display_name", "Baggings, Bilbo LWRS:EX");
```
Comment (craigyu) [inline backend/src/test/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelperTest.java:43]: nice 🧙‍♂️

**Review by craigyu — CHANGES_REQUESTED** [review]
Great work! I think a userId is needed for the User object, we are storing userId only and that's the only way we can distinguish whether the user is using idir or bceid on the DB

File: backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java [hunk 3]
```diff
@@ -31,42 +26,58 @@ public Optional<UserInfo> getUserInfo() {
...
         }
 
+        // Provider IDIR or BCeID & username
+        String provider = jwtPrincipal.getClaimAsString("custom:idp_name");
+        boolean isIdirProvider = provider.equals("idir");
+        String idpUsername = jwtPrincipal.getClaimAsString("custom:idp_username");
+
+        // User name
+        String displayName =
+            jwtPrincipal.getClaimAsString("custom:idp_display_name"); // de Campos, Ricardo WLRS:EX
```
Comment (mgaseta) [inline backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java:42]: Suggestion (non-blocking): for extra security, maybe it's better to remove real data from the comments.

File: backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java [hunk 4]
```diff
@@ -31,42 +26,58 @@ public Optional<UserInfo> getUserInfo() {
...
+        String firstName = "";
+        String lastName = "";
+
+        // Usually only IDIR contains comma. E.g.: de Campos, Ricardo WLRS:EX
```
Comment (mgaseta) [inline backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java:46]: Suggestion (non-blocking): for extra security, maybe it's better to remove real data from the comments.

File: backend/src/test/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelperTest.java [hunk 5]
```diff
@@ -76,26 +70,26 @@ void getUserInfoBusinessBceidTest() {
...
     Jwt.Builder builder = Jwt.withTokenValue("myTokenValue");
-    builder.subject("mordor@bceid");
+    builder.subject("MORDOR-BCEID");
     builder.header("alg", "HS256");
     builder.header("typ", "JWT");
     builder.claim("email", "lord.sauron@mordor.middleearth");
-    builder.claim("display_name", "Sauron, Mordor LWRS:EX");
-    builder.claim("bceid_business_name", "Mordor LLC");
-    builder.claim("identity_provider", "bceidbusiness");
+    builder.claim("custom:idp_display_name", "Lord Sauron of Mordor");
```
Comment (mgaseta) [inline backend/src/test/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelperTest.java:77]: Sauron is watching.

giphy

**Review by mgaseta — APPROVED** [review]
Really nice work! :100: :1st_place_medal: 
Besides Craig's comment, I added only a small non-blocking suggestion.

File: backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java (hunk 1 above)
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java]: Done here! https://github.com/bcgov/nr-spar/pull/585/commits/7ea3008c687b03face7611aa2f6b221d3a28ec61

File: backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java (hunk 3 above)
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java]: Nice! Thanks. Done! https://github.com/bcgov/nr-spar/pull/585/commits/24d2ce531bc59094974b3b68c0465d7bbb6ab278

File: backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java (hunk 4 above)
Comment (RMCampos) [reply backend/src/main/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelper.java]: Done here! https://github.com/bcgov/nr-spar/pull/585/commits/15bdb6974bb396367e08d8231671d1396b8f8db3

File: backend/src/test/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelperTest.java (hunk 5 above)
Comment (RMCampos) [reply backend/src/test/java/ca/bc/gov/backendstartapi/security/UserAuthenticationHelperTest.java:78]: I hope he is! We're getting there 🧙🏻‍♂️

**Review by craigyu — APPROVED** [review]

**Pull request merged**